docs/control-plane-architecture.adoc

```diff
@@ -0,0 +1,252 @@
+
+Service discovery happens via DNS.  See <<rfd206>> and <<rfd248>>.
+
+=== Nexus, data flow
+
+Nexus is the place where system-wide decisions get made.  CockroachDB is the source of truth for all configuration.  
+
+Nexus stores all of its state in CockroachDB.  It's the only component that communicates directly with CockroachDB.
+
+Nexus instances operate independently, without directly coordinating with each other except through CockroachDB.
+
+Generally, when a change gets made, the process is:
+
+1. Nexus receives a request to make the change (e.g., via the external API)
+2. Nexus validates the requested change
+3. Nexus stores the information into CockroachDB.  (This is the point where change is serialized against any concurrent changes.)
+4. Nexus propagates the change to other components that need to know about it.
+
+There are a few basic contexts in Nexus:
+
+* **API requests** from either the external or internal API.  Here, Nexus is latency-sensitive.  When we make database queries or other requests in this context, we usually do _not_ retry transient failures, but leave that to callers (See https://en.wikipedia.org/wiki/End-to-end_principle["end-to-end principle"]).  API request handlers may kick off sagas or activate background tasks.
+* **Distributed sagas** are a https://www.youtube.com/watch?v=0UTOLRTwOX0[design pattern] for carrying out multi-step operations in a distributed system.  Saga actions generally _do_ retry transient errors indefinitely.
+* **Background tasks** are periodic or event-triggered activities that manage everything else that has to happen in the system (e.g., change propagation, CockroachDB cluster management, fault tolerance, etc.).  Nexus has a framework for background tasks that's oriented around the "reconciler" pattern (see <<rfd373>>).  In this context, we also usually don't retry individual operations -- instead, the entire activity will be retried on a periodic basis.  Background tasks are structured to re-evaluate the state of the world each time they're run and then determine what to do, on the assumption that things may have changed since the last time they ran.
+
+It's essential that components provide visibility into what they're doing for debugging and support.  ("Software should be able to exonerate itself when things are broken.")
```

Comment:
What is this quote from?

Reply:
Um, I think it's just something I've said a lot. 😄 I'll remove the quotes so it looks less like a reference.